Support border/margin/padding on msub/msup/msubsup

fred-wang opened the pull request:
#### df1a1c2be0c95f797e04301be8b8471af5176537
<pre>
Support border/margin/padding on msub/msup/msubsup
<a href="https://bugs.webkit.org/show_bug.cgi?id=276314">https://bugs.webkit.org/show_bug.cgi?id=276314</a>

Reviewed by Rob Buis.

This implements support for border/margin/padding on the msub, msup
and msubsup elements according to the rules from MathML Core:

- When handling boxes of children during math layout, we consider *margin*
  boxes. `recomputeLogicalWidth()` is already called during the layout
  of children to set the inline margins (this does not seem enough for
  preferred width calculations though, but will be considered in
  follow-up patches). This patchs adds a logic to call
  `computeAndSetBlockDirectionMargins()` on each child in order to set
  their block margins.

- Current math layout is modified so that padding/border are added around
  the current layout in order to obtain the border box.

Canonical link: <a href="https://commits.webkit.org/280768@main">https://commits.webkit.org/280768@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/501814ebfe3e776e280a2bcebfe61faa462b30e4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57575 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36903 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10050 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61197 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8020 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59703 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44536 "Hash 501814eb for PR 30565 does not build (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8208 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46616 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5687 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59605 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/44536 "Hash 501814eb for PR 30565 does not build (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49743 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27482 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/44536 "Hash 501814eb for PR 30565 does not build (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7041 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7023 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/44536 "Hash 501814eb for PR 30565 does not build (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7313 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62877 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1488 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7401 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53876 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1495 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49762 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53979 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12732 "Failed to checkout and rebase branch from PR 30565") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1264 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8587 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32732 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33817 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34902 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33563 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis reviewed (approved):
LGTM. I think "This patchs adds a call to call" -> "This patch adds logic to call"

fred-wang committed:
Support border/margin/padding on msub/msup/msubsup
https://bugs.webkit.org/show_bug.cgi?id=276314

Reviewed by Rob Buis.

This implements support for border/margin/padding on the msub, msup
and msubsup elements according to the rules from MathML Core:

- When handling boxes of children during math layout, we consider *margin*
  boxes. `recomputeLogicalWidth()` is already called during the layout
  of children to set the inline margins (this does not seem enough for
  preferred width calculations though, but will be considered in
  follow-up patches). This patchs adds a logic to call
  `computeAndSetBlockDirectionMargins()` on each child in order to set
  their block margins.

- Current math layout is modified so that padding/border are added around
  the current layout in order to obtain the border box.

Canonical link: https://commits.webkit.org/280768@main
